# Build shared library and fix tests run for it

diff -N -r -U 3 ocrad-0.20-orig/Makefile.in ocrad-0.20/Makefile.in
--- ocrad-0.20-orig/Makefile.in	2010-07-14 17:53:06.000000000 +0300
+++ ocrad-0.20/Makefile.in	2010-08-30 13:35:59.000000000 +0300
@@ -13,26 +13,26 @@
            bitmap.o blob.o profile.o feats.o feats_test0.o feats_test1.o \
            character.o character_r11.o character_r12.o character_r13.o \
            textline.o textline_r2.o textblock.o textpage.o
-objs     = arg_parser.o main.o
+prg_objs = arg_parser.o main.o
 
 
 .PHONY : all install install-info install-man install-strip \
          uninstall uninstall-info uninstall-man \
          doc info man check dist clean distclean
 
-all : $(progname) lib$(libname).a
+all : lib$(libname).so $(progname)
 
-lib$(libname).a: $(ocr_objs) $(lib_objs)
-	$(AR) -rcs lib$(libname).a $(ocr_objs) $(lib_objs)
+lib$(libname).so: $(ocr_objs) $(lib_objs)
+	$(CXX) -shared -Wl,-soname,lib$(libname).so -o lib$(libname).so $(ocr_objs) $(lib_objs)
 
-$(progname) : $(ocr_objs) $(objs)
-	$(CXX) $(LDFLAGS) -o $(progname) $(ocr_objs) $(objs)
+$(progname) : lib$(libname).so $(prg_objs)
+	$(CXX) $(LDFLAGS) -o $(progname) -L. -locrad $(prg_objs)
 
-$(progname)_profiled : $(ocr_objs) $(objs)
-	$(CXX) $(LDFLAGS) -pg -o $(progname)_profiled $(ocr_objs) $(objs)
+$(progname)_profiled : lib$(libname).so $(prg_objs)
+	$(CXX) $(LDFLAGS) -pg -o $(progname)_profiled -L. -locrad $(prg_objs)
 
-ocrcheck : ocrcheck.o lib$(libname).a
-	$(CXX) $(LDFLAGS) -o ocrcheck ocrcheck.o lib$(libname).a
+ocrcheck : ocrcheck.o lib$(libname).so
+	$(CXX) $(LDFLAGS) -o ocrcheck -L. -locrad ocrcheck.o
 
 ocrcheck.o : ocrcheck.cc
 	$(CXX) $(CPPFLAGS) $(CXXFLAGS) -DPROGVERSION=\"$(pkgversion)\" -c -o $@ $<
@@ -41,11 +41,11 @@
 	$(CXX) $(CPPFLAGS) $(CXXFLAGS) -DPROGVERSION=\"$(pkgversion)\" -c -o $@ $<
 
 %.o : %.cc
-	$(CXX) $(CPPFLAGS) $(CXXFLAGS) -c -o $@ $<
+	$(CXX) $(CPPFLAGS) $(CXXFLAGS) -fPIC -c -o $@ $<
 
 $(lib_objs)     : Makefile ocradlib.h
 $(ocr_objs)     : Makefile bitmap.h blob.h common.h rational.h rectangle.h ucs.h
-$(objs)         : Makefile arg_parser.h
+$(prg_objs)     : Makefile arg_parser.h
 character.o     : character.h
 character_r11.o : character.h profile.h feats.h
 character_r12.o : character.h profile.h feats.h
@@ -84,7 +84,7 @@
 	./config.status
 
 check : all ocrcheck
-	@$(VPATH)/testsuite/check.sh $(VPATH)/testsuite
+	$(VPATH)/testsuite/check.sh $(VPATH)/testsuite
 
 install : all install-info install-man
 	if [ ! -d "$(DESTDIR)$(bindir)" ] ; then $(INSTALL_DIR) "$(DESTDIR)$(bindir)" ; fi
@@ -92,7 +92,7 @@
 	if [ ! -d "$(DESTDIR)$(libdir)" ] ; then $(INSTALL_DIR) "$(DESTDIR)$(libdir)" ; fi
 	$(INSTALL_PROGRAM) ./$(progname) "$(DESTDIR)$(bindir)/$(progname)"
 	$(INSTALL_DATA) $(VPATH)/$(libname)lib.h "$(DESTDIR)$(includedir)/$(libname)lib.h"
-	$(INSTALL_DATA) ./lib$(libname).a "$(DESTDIR)$(libdir)/lib$(libname).a"
+	$(INSTALL_DATA) ./lib$(libname).so "$(DESTDIR)$(libdir)/lib$(libname).so"
 
 install-info :
 	if [ ! -d "$(DESTDIR)$(infodir)" ] ; then $(INSTALL_DIR) "$(DESTDIR)$(infodir)" ; fi
@@ -109,7 +109,7 @@
 uninstall : uninstall-info uninstall-man
 	-rm -f "$(DESTDIR)$(bindir)/$(progname)"
 	-rm -f "$(DESTDIR)$(includedir)/$(libname)lib.h"
-	-rm -f "$(DESTDIR)$(libdir)/lib$(libname).a"
+	-rm -f "$(DESTDIR)$(libdir)/lib$(libname).so"
 
 uninstall-info :
 	-install-info --info-dir="$(DESTDIR)$(infodir)" --remove "$(DESTDIR)$(infodir)/$(pkgname).info"
@@ -143,8 +143,8 @@
 	lzip -v -9 $(DISTNAME).tar
 
 clean :
-	-rm -f $(progname) $(progname)_profiled $(objs) $(ocr_objs)
-	-rm -f ocrcheck ocrcheck.o $(lib_objs) *.a
+	-rm -f $(progname) $(progname)_profiled $(prg_objs) $(ocr_objs)
+	-rm -f ocrcheck ocrcheck.o $(lib_objs) *.a *.so
 
 distclean : clean
 	-rm -f Makefile config.status *.tar *.tar.lz
diff -N -r -U 3 ocrad-0.20-orig/testsuite/check.sh ocrad-0.20/testsuite/check.sh
--- ocrad-0.20-orig/testsuite/check.sh	2010-07-15 13:45:34.000000000 +0300
+++ ocrad-0.20/testsuite/check.sh	2010-08-30 13:38:10.000000000 +0300
@@ -5,8 +5,8 @@
 # This script is free software: you have unlimited permission
 # to copy, distribute and modify it.
 
-LC_ALL=C
-export LC_ALL
+export LC_ALL=C
+export LD_LIBRARY_PATH="$LD_LIBRARY_PATH;.;.."
 objdir=`pwd`
 testdir=`cd "$1" ; pwd`
 OCRAD="${objdir}"/ocrad
